File: src/main/rs/emulate/shared/property/PropertyDecoders.java
package rs.emulate.shared.property;

import java.util.function.Function;

import rs.emulate.shared.util.DataBuffer;

/**
 * Contains static {@link Function}s to decode values.
 * 
 * @author Major
 */
public final class PropertyDecoders {

	/**
	 * A {@link Function} that decodes an unsigned {@code byte} from the specified {@link DataBuffer}.
	 */
	public static final Function<DataBuffer, Integer> BYTE_DECODER = DataBuffer::getUnsignedByte;

	/**
	 * A {@link Function} that reads no data from the {@link DataBuffer} and simply returns {@code false}.
	 */
	public static final Function<DataBuffer, Boolean> FALSE_DECODER = buffer -> false;

	/**
	 * A {@link Function} that decodes an unsigned {@code int} from the specified {@link DataBuffer}.
	 */
	public static final Function<DataBuffer, Integer> INT_DECODER = DataBuffer::getInt;

	/**
	 * A {@link Function} that decodes an unsigned {@code short} from the specified {@link DataBuffer}.
	 */
	public static final Function<DataBuffer, Integer> SHORT_DECODER = DataBuffer::getUnsignedShort;

	/**
	 * A {@link Function} that decodes a signed {@code byte} from the specified {@link DataBuffer}.
	 */
	public static final Function<DataBuffer, Integer> SIGNED_BYTE_DECODER = DataBuffer::getByte;

	/**
	 * A {@link Function} that decodes a signed {@code short} from the specified {@link DataBuffer}.
	 */
	public static final Function<DataBuffer, Integer> SIGNED_SHORT_DECODER = DataBuffer::getShort;

	/**
	 * A {@link Function} that decodes a smart from the specified {@link DataBuffer}.
	 */
	public static final Function<DataBuffer, Integer> SMART_DECODER = DataBuffer::getSmart;

	/**
	 * A {@link Function} that decodes an old-style (10-terminated) String from the specified {@link DataBuffer}.
	 */
	public static final Function<DataBuffer, String> ASCII_STRING_DECODER = DataBuffer::getString;

	/**
	 * A {@link Function} that decodes an unsigned {@code tri-byte} from the specified {@link DataBuffer}.
	 */
	public static final Function<DataBuffer, Integer> TRI_BYTE_DECODER = DataBuffer::getUnsignedTriByte;

	/**
	 * A {@link Function} that reads no data from the {@link DataBuffer} and simply returns {@code true}.
	 */
	public static final Function<DataBuffer, Boolean> TRUE_DECODER = buffer -> true;

	/**
	 * Sole private constructor to prevent instantiation.
	 */
	private PropertyDecoders() {

	}

}

File: util/src/main/kotlin/rs/emulate/shared/property/PropertyDecoders.kt
package rs.emulate.shared.property

import rs.emulate.shared.util.DataBuffer

import java.util.function.Function

/**
 * Contains static [Function]s to decode values.
 */
object PropertyDecoders {

    val ASCII_STRING_DECODER: (DataBuffer) -> String = DataBuffer::getString

    val BYTE_DECODER: (DataBuffer) -> Int = DataBuffer::getUnsignedByte

    val FALSE_DECODER: (DataBuffer) -> Boolean = { false }

    val INT_DECODER: (DataBuffer) -> Long = DataBuffer::getUnsignedInt

    val SHORT_DECODER: (DataBuffer) -> Int = DataBuffer::getUnsignedShort

    val SIGNED_BYTE_DECODER: (DataBuffer) -> Int = DataBuffer::getByte

    val SIGNED_INT_DECODER: (DataBuffer) -> Int = DataBuffer::getInt

    val SIGNED_SHORT_DECODER: (DataBuffer) -> Int = DataBuffer::getShort

    val SMART_DECODER: (DataBuffer) -> Int = DataBuffer::getSmart

    val TRUE_DECODER: (DataBuffer) -> Boolean = { true }

    val UNSIGNED_TRI_BYTE_DECODER: (DataBuffer) -> Int = DataBuffer::getUnsignedTriByte

}
